src/compiler/checker.ts

```diff
@@ -20244,18 +20244,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const typeArguments = map(typeParameters, t => getMappedType(t, combinedMapper));
             const newAliasSymbol = aliasSymbol || type.aliasSymbol;
             const newAliasTypeArguments = aliasSymbol ? aliasTypeArguments : instantiateTypes(type.aliasTypeArguments, mapper);
-            const id = (type.objectFlags & ObjectFlags.SingleSignatureType ? "S" : "") + getTypeListId(typeArguments) + getAliasId(newAliasSymbol, newAliasTypeArguments);
+            const id = getTypeListId(typeArguments) + getAliasId(newAliasSymbol, newAliasTypeArguments);
             if (!target.instantiations) {
                 target.instantiations = new Map<string, Type>();
                 target.instantiations.set(getTypeListId(typeParameters) + getAliasId(target.aliasSymbol, target.aliasTypeArguments), target);
             }
             let result = target.instantiations.get(id);
             if (!result) {
-                if (type.objectFlags & ObjectFlags.SingleSignatureType) {
-                    result = instantiateAnonymousType(type, mapper);
-                    target.instantiations.set(id, result);
-                    return result;
-                }

```

Comment:
It might feel like a downgrade that those are not cached right now. `getObjectTypeInstantiation` (the containing function) is only called from a single place and all the functions called from the branch leading to `getObjectTypeInstantiation` are implementing instantiation caching (`createNormalizedTypeReference`, `instantiateReverseMappedType` and `getObjectTypeInstantiation`).

But now not the branch I added - the one that calls `instantiateAnonymousType` for single signature types. I have no idea what I would use to create the cache key there. The only unique bits there are the results of the `mapper` but `mapper` isn't exactly called eagerly there so it doesn't have access to those.

However, I have a feeling that perhaps caching is already - sort of - implemented for those single signature types. It's just at a different level, like in the `getReturnTypeOfSignature`.

Reply:
My gut is making me worried that we're dropping some perf here due to not caching quickly, but maybe it doesn't matter at all?